evaluation/TUTORIAL.md

```diff
@@ -0,0 +1,517 @@
+### Before everything begins
+
+Please follow [this document](https://github.com/OpenDevin/OpenDevin/blob/main/Development.md) to setup local develop environment for OpenDevin.
+
+### Configuration file
+
+OpenDevin uses `config.toml` to keep track of most configurations.
+
+Here's an example configuration file you can use:
+
+```toml
+[core]
+max_iterations = 100
+cache_dir = "/tmp/cache"
+
+# IMPORTANT: You should set these two paths to YOUR WORKSPACE directory,
+# which will be mounted into Sandbox for agent to interact with!
+# The OpenDevin agent will be able to read/write files whatever they like (even rm -rf)
+# in this directory, so be careful!!
+workspace_base = "/home/xingyaow/OpenDevin-dev/workspace"
+workspace_mount_path = "/home/xingyaow/OpenDevin-dev/workspace"
+# ==========================
+
+sandbox_container_image = "ghcr.io/opendevin/sandbox:latest"
+run_as_devin = true
+sandbox_type = "ssh"
+use_host_network = true
+ssh_hostname = "localhost"
+sandbox_timeout = 120
+
+[llm]
+model = "gpt-4o-2024-05-13"
+api_key = "sk-XXX"
+```
+
+### How do OpenDevin solves a task?
+
+In this section, for the purpose of building an evaluation task, we omit details on the frontend and backend that mostly contains logic for user interface, and mainly focus on a function `main` (i.e., a command line interface) that allows you to complete a task end-to-end.
+
+For example, *without any frontend/backend server*, you can run ([setup the environment first](#before-everything-begins) if you haven't) the following to get a simple task solved (with max number of 10 iterations, `CodeActAgent` as the agent, and `gpt-4o-2024-05-13` as the LLM):
+
+```bash
+poetry run python ./opendevin/core/main.py \
+        -i 10 \
+        -t "Write me a bash script that print hello world." \
+        -c CodeActAgent \
+        -m gpt-4o-2024-05-13
+```
+
+After running the script, you will observe the following:
+
+![](./static/example_task_1.png)
+
+You can see the agent uses bash to write a script, making it executable, and then tested it by running it to make sure it is working.
+
+## What is "Request user input"?
+
+At the end of the above screenshot, OpenDevin actually requests user inputs when it think it finishes the task. There are multiple motivations for this design:
+
+1. If the agent did something wrong, the user can follow-up with language feedback to correct it.
+
+2. If the agent did something good, the user may want to follow-up with additional problem for it to solve.
+
+3. If the agent completes the user's task -- in a realistic setting, the user might just turn off OpenDevin application and work on something else. User could also type in "/exit" to gracefully shutdown the OpenDevin agent.
+
+However, this will actually causes issues in evaluation, since most evaluation don't actually assume additional user input.
+
+To fix this, we introduce the functionality of `fake_user_response_fn` in the `main` fucntion. Check the section below for more details!
```

Comment:
```suggestion
At the end of the above screenshot, OpenDevin actually requests user inputs when it think it finishes the task. This will cause issues in evaluation, since most evaluation don't assume additional user input. To fix this, we introduce the functionality of `fake_user_response_fn` in the `main` function, which we describe in the next section.
```